test/common/router/retry_state_impl_test.cc

```diff
@@ -412,6 +412,84 @@ TEST_F(RouterRetryStateImplTest, RetriableStatusCodesHeader) {
   }
 }
 
+// Test that when 'retriable-headers' policy is set via request header, certain configured headers
+// trigger retries.
+TEST_F(RouterRetryStateImplTest, RetriableHeadersPolicySetViaRequestHeader) {
+  policy_.retriable_headers_.emplace_back("x-upstream-pushback");
+  Http::TestHeaderMapImpl request_headers{{"x-envoy-retry-on", "retriable-headers"}};
+  setup(request_headers);
+  EXPECT_TRUE(state_->enabled());
+
+  expectTimerCreateAndEnable();
+
+  Http::TestHeaderMapImpl response_headers{{":status", "200"}, {"x-upstream-pushback", "true"}};
+  EXPECT_EQ(RetryStatus::Yes, state_->shouldRetryHeaders(response_headers, callback_));
+}
+
+// Test that when 'retriable-headers' policy is set via retry policy configuration, certain
+// configured headers trigger retries.
+TEST_F(RouterRetryStateImplTest, RetriableHeadersPolicyViaRetryPolicyConfiguration) {
+  policy_.retry_on_ = RetryPolicy::RETRY_ON_RETRIABLE_HEADERS;
+  policy_.retriable_headers_.emplace_back("x-upstream-pushback");
+  Http::TestHeaderMapImpl request_headers;
+  setup(request_headers);
+  EXPECT_TRUE(state_->enabled());
+
+  expectTimerCreateAndEnable();
+
+  Http::TestHeaderMapImpl response_headers{{":status", "200"}, {"x-upstream-pushback", "true"}};
+  EXPECT_EQ(RetryStatus::Yes, state_->shouldRetryHeaders(response_headers, callback_));
+}
+
+// Test that when 'retriable-headers' policy is not set, configured retrieable headers are ignored.
```

Comment:
The spell checker tagged "retrieable". The other test failure looks like a flake.

Also, we prefer not to have force-pushed changes since it can mess up github's comment tracking. We'll squash the commits on merge so you don't have to worry about keeping the history clean.